### PR TITLE
Add `dnsServers` to `ProviderSpec`

### LINKED ISSUE #91: Make DNS server configurable in `MachineClass`
# Summary

Extend the existing `ProviderSpec` with a `dnsServers` field which allows the configuration of custom DNS servers. The result of this should be rendered into the `Machine` ignition to configure the `resolvd`:

```
    - path: /etc/systemd/resolved.conf.d/dns.conf
      mode: 0644
      contents:
        inline: |
          [Resolve]
          DNS=x.x.x.x
          DNS=y.y.y.y
```

## Definition of done

- Extend the `ProviderSpec` with a `dnsServers` field
- Add validation to ensure the list of DNS servers are valid IP addresses
- Ensure the correct behaviour by providing the corresponding unit tests

### PR DESCRIPTION
**Pull request by ushabelgur**
# Proposed Changes
- Extended the ProviderSpec with a dnsServers field
- Added validation to ensure the list of DNS servers are valid IP addresses
- Added corresponding unit tests

Fixes #91

### CODE REVIEW

File: pkg/ignition/config.go [hunk 1]
```diff
@@ -67,6 +76,31 @@ func File(config *Config) (string, error) {
 			return "", err
 		}
 	}
+	if len(config.DnsServers) > 0 {
```
Comment (afritzler) [inline pkg/ignition/config.go:79]: No need to check for 0 here as you are having the for loop below.

File: pkg/ignition/config.go [hunk 2]
```diff
@@ -67,6 +76,31 @@ func File(config *Config) (string, error) {
...
+					"mode": fileMode,
+					"contents": map[string]interface{}{
+						"inline": strings.Join(dnsServers, "\n"),
+					},
+				}},
+			},
+		}
+
+		// merge dnsConfiguration with ignition content
+		err := mergo.Merge(ignitionBase, dnsConf, mergo.WithAppendSlice)
```
Comment (afritzler) [inline pkg/ignition/config.go:99]: Please use `if err := ... ; err != nil` in one line here.

File: pkg/ignition/config.go [hunk 3]
```diff
@@ -67,6 +76,31 @@ func File(config *Config) (string, error) {
...
+		if err != nil {
+			return "", err
```
Comment (afritzler) [inline pkg/ignition/config.go:101]: Please return an `fmt.Errorf("some meaningful error message: %w", err)` here.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/ignition/config.go (hunk 1 above)
Comment (ushabelgur) [reply pkg/ignition/config.go:79]: in for loop i am just combining all the IPs, if i remove this check default content will get added to ignition content without IPs like this 
 - path: /etc/systemd/resolved.conf.d/dns.conf
      mode: 0644
      contents:
        inline: |
          [Resolve]
 Or i can move this check before forming the dnsConf content at line no. 86

File: pkg/api/validation/validation.go [hunk 4]
```diff
@@ -64,5 +67,11 @@ func validateOnmetalMachineClassSpec(spec *v1alpha1.ProviderSpec, fldPath *field
 		allErrs = append(allErrs, field.Required(fldPath.Child("prefixname"), "prefixname is required"))
 	}
 
+	for _, ip := range spec.DnsServers {
+		if !netip.Addr.IsValid(ip) {
+			allErrs = append(allErrs, field.Required(fldPath.Child("dnsserver"), "dnsserver ip is invalid"))
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/api/validation/validation.go]: Done

File: pkg/ignition/config.go (hunk 2 above)
Comment (ushabelgur) [reply pkg/ignition/config.go]: Done

File: pkg/ignition/config.go (hunk 3 above)
Comment (ushabelgur) [reply pkg/ignition/config.go]: Done

File: pkg/ignition/config.go [hunk 5]
```diff
@@ -67,6 +76,28 @@ func File(config *Config) (string, error) {
...
+					"inline": strings.Join(dnsServers, "\n"),
+				},
+			}},
+		},
+	}
+
+	// merge dnsConfiguration with ignition content
+	if err := mergo.Merge(ignitionBase, dnsConf, mergo.WithAppendSlice); err != nil {
+		return "", fmt.Errorf("failed to merge dnsServer configuration with igntition content: %w", err)
+	}
```
Comment (afritzler) [inline pkg/ignition/config.go:100]: You are right having a check for `dnsServer` `len` here makes sense. We don't want to allocate anything if no `dnsServers` are set.

**Review by afritzler — CHANGES_REQUESTED** [review]

**Comment by afritzler** [comment]
The only thing we can actually validate is:
```
// IsValid reports whether the Addr is an initialized address (not the zero Addr).
//
// Note that "0.0.0.0" and "::" are both valid values.
func (ip Addr) IsValid() bool { return ip.z != z0 }
```
If the `Addr` is non zero `Addr`. A zero address basically means:

```
z0    = (*intern.Value)(nil)
```

Therefore the only thing we need to do in the test is to ensure that the validation fails if we get a zero address.

**Comment by ushabelgur** [comment]
The zero Addr is not a valid IP address. Addr{} is distinct from both 0.0.0.0 and ::.
package main

import (
	"fmt"
	"net/netip"
)

func main() {
	fmt.Printf("%#v", netip.Addr{})
}
Output: netip.Addr{addr:netip.uint128{hi:0x0, lo:0x0}, z:(*intern.Value)(nil)}

File: pkg/api/v1alpha1/provider.go [hunk 6]
```diff
@@ -45,6 +47,9 @@ type ProviderSpec struct {
 	PrefixName string `json:"prefixName"`
 	// Labels are used to tag resources which the MCM creates, so they can be identified later.
 	Labels map[string]string `json:"labels,omitempty"`
+	// dnsServers is list of IPs which allows to configure the custom DNS servers.
+	// The result of this will be rendered into the Machine ignition to configure the resolvd
```
Comment (afritzler) [inline pkg/api/v1alpha1/provider.go:51]:
```suggestion
	// DnsServers is a list of DNS resolvers which should be configured on the host.
```

Comment (ushabelgur) [reply pkg/api/v1alpha1/provider.go]: Done

File: pkg/onmetal/create_machine_test.go [hunk 7]
```diff
@@ -161,5 +162,15 @@ var _ = Describe("CreateMachine", func() {
...
 		}).Should(Succeed())
+
+		By("failing if the invalid IP is set for dnsServers in provider")
+		Eventually(func(g Gomega) {
+			_, err := (*drv).CreateMachine(ctx, &driver.CreateMachineRequest{
+				Machine:      newMachine(ns, "machine", -1, nil),
+				MachineClass: newMachineClass(v1alpha1.ProviderName, internal.ProviderSpecWithInvalidDNS),
+				Secret:       providerSecret,
+			})
+			g.Expect(err.Error()).To(ContainSubstring("unable to parse IP]"))
```
Comment (afritzler) [inline pkg/onmetal/create_machine_test.go:173]: Please remove the bracket at the end.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/onmetal/create_machine_test.go (hunk 7 above)
Comment (ushabelgur) [reply pkg/onmetal/create_machine_test.go]: Done

**Review by afritzler — APPROVED** [review]
